### PR TITLE
Checking for immutability solution v2

### PR DESCRIPTION
**Pull request by andredered**
Signed-off-by: Andrii Derevianko <andredered@gmail.com>

### CODE REVIEW

File: ArchUnitNETTests/Fluent/Extensions/BuildMocksExtensions.cs [hunk 1]
```diff
@@ -130,6 +130,18 @@ public static MethodMember CreateStubMethodMember(this MethodBase methodBase)
             var parameters = methodBase.CreateStubParameters().Select(parameter => new TypeInstance<IType>(parameter));
             var methodForm = methodBase.GetStubMethodForm();
             var isStatic = methodBase.IsStatic;
+            
+            //TODO isReadOnly = ?
```
Comment (fgathercoupa) [inline ArchUnitNETTests/Fluent/Extensions/BuildMocksExtensions.cs:134]: I actually wonder, why this class is in Fluent and not in domain @fholger

File: global.json [hunk 2]
```diff
@@ -1,5 +1,7 @@
 {
   "sdk": {
-    "version": "5.0.203"
+    "version": "5.0.203",
+    "rollForward": "latestMajor",
```
Comment (fholger) [inline global.json:4]: What are these changes for?

**Review by fholger — COMMENTED** [review]
In general, it looks good to me. But we should add some tests to make sure it actually works as intended:
- in ArchUnitNETTests/Domain/MemberTests we should add tests to see that `IsReadOnly` gets the proper value. Would need a test class that has e.g. `readonly` fields, properties with only `{get;}`, properties with `{get;init;}` and properties with `{get;set;}`
- in ArchUnitNETTests/Fluent/Syntax/Elements/MemberSyntaxElementsTests we should have basic tests for the fluent accessors, similar to the tests for `Static` that are already in there. But we need to ensure that there are classes with relevant read-only members in the test architecture so that these tests actually do anything.

**Review by fholger — COMMENTED** [review]
A general concern I have is that it feels to me like we are mixing read-only and init-only properties together under the `IsReadOnly` label, but they are technically not the same. Actually I think it would be more appropriate to have a three-state enum `Writability` with values `ReadOnly`, `InitOnly`and `Writable` instead.

But perhaps that warrants some follow-up discussion and should be done in another PR. I think we can probably merge this one as a basis to work from.

File: ArchUnitNET/Domain/FieldMember.cs [hunk 3]
```diff
@@ -13,9 +13,9 @@ namespace ArchUnitNET.Domain
     public class FieldMember : IMember, ITypeInstance<IType>
     {
         private readonly ITypeInstance<IType> _typeInstance;
-
+
```
Comment (MSinstein) [inline ArchUnitNET/Domain/FieldMember.cs:16]: Please revert this whitespace change

File: ArchUnitNET/Fluent/Syntax/Elements/Members/GivenMembersThat.cs [hunk 4]
```diff
@@ -71,6 +71,11 @@ public TGivenRuleTypeConjunction AreStatic()
             return Create<TGivenRuleTypeConjunction, TRuleType>(_ruleCreator);
         }
 
+        public TGivenRuleTypeConjunction AreReadOnly()
+        {
+            _ruleCreator.AddPredicate(MemberPredicatesDefinition<TRuleType>.AreReadOnly());
+            return Create<TGivenRuleTypeConjunction, TRuleType>(_ruleCreator);
+        }
         //Negations
```
Comment (MSinstein) [inline ArchUnitNET/Fluent/Syntax/Elements/Members/GivenMembersThat.cs:79]: This comment belongs to the next section. Please move it down a little, e.g. by adding a new line after the new method

File: ArchUnitNET/Fluent/Syntax/Elements/Members/MemberConditionsDefinition.cs [hunk 5]
```diff
@@ -361,8 +367,12 @@ public static ICondition<TRuleType> NotBeStatic()
...
         }
-
-
+        
+        public static ICondition<TRuleType> NotBeReadOnly()
+        {
+            return new SimpleCondition<TRuleType>(member => !member.IsReadOnly.HasValue || !member.IsReadOnly.Value,
+                "not be read only", "is read only");
+        }
         //Relation Condition Negations
```
Comment (MSinstein) [inline ArchUnitNET/Fluent/Syntax/Elements/Members/MemberConditionsDefinition.cs:376]: same here

File: ArchUnitNET/Loader/TypeFactory.cs [hunk 6]
```diff
@@ -307,6 +311,7 @@ private ITypeInstance<IType> CreateTypeFromTypeReference(TypeReference typeRefer
                 isIterator = methodDefinition.IsIterator();
                 isStatic = methodDefinition.IsStatic;
                 isStub = false;
+
```
Comment (MSinstein) [inline ArchUnitNET/Loader/TypeFactory.cs:314]: please remove this newline again

File: ArchUnitNETTests/Domain/MemberTests.cs [hunk 7]
```diff
@@ -257,5 +274,17 @@ private FieldType FieldWithoutGetter
         public class FieldType
         {
         }
+        private class ClassReadOnly
+        {
+            private readonly string readonlyVar;
+            private readonly string readonlyVarInit = "";
+            private string NotReadOnlyVarInit = "";
+            private string PropertyWithoutSet { get; }
+            private string PropertyWithGetUndSet { get; set; }
```
Comment (MSinstein) [inline ArchUnitNETTests/Domain/MemberTests.cs:283]: typo
```suggestion
            private string PropertyWithGetAndSet { get; set; }
```

File: ArchUnitNETTests/Domain/MemberTests.cs [hunk 8]
```diff
@@ -188,7 +204,7 @@ public void MembersAreMethodMembersAndFieldMembersAndPropertyMembers()
             Assert.True(fieldMembers.All(member => members.Contains(member)));
             Assert.True(propertyMembers.All(member => members.Contains(member)));
         }
-
+
```
Comment (MSinstein) [inline ArchUnitNETTests/Domain/MemberTests.cs:207]: please revert this whitespace change

**Pull request merged**